src/compiler/transformers/declarations/localInferenceResolver.ts

```diff
@@ -0,0 +1,879 @@
+                return factory.createNumericLiteral(+literal.text);
+        }
+        throw new Error("Not supported");
+    }
+    function createUndefinedTypeNode() {
+        if (strictNullChecks) {
+            return factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword);
+        }
+        else {
+            return factory.createKeywordTypeNode(SyntaxKind.AnyKeyword);
+        }
+    }
+    function literal(node: Node, baseType: string | KeywordTypeSyntaxKind, narrowBehavior: NarrowBehavior) {
+        if (narrowBehavior & NarrowBehavior.AsConstOrKeepLiterals) {
+            return factory.createLiteralTypeNode(
+                normalizeLiteralValue(node as LiteralExpression),
+            );
+        }
+        else {
+            return typeof baseType === "number" ? factory.createKeywordTypeNode(baseType) : factory.createTypeReferenceNode(baseType);
+        }
+    }
+
+    function visitTypeAndClone(type: TypeNode) {
+        const visitedType = visitNode(type, visitDeclarationSubtree, isTypeNode)!;
+        return deepClone(visitedType);
+    }
+
+    function normalizePropertyName(symbol: Symbol, isMethod: boolean) {
+        let nameText;
+        Debug.assert(symbol.declarations !== undefined, "Symbol has no declarations");
+        let stringNamed = !!length(symbol.declarations);
+        let singleQuote = stringNamed;
+        for (const declaration of symbol.declarations) {
+            const name = getNameOfDeclaration(declaration);
+            if (!name) {
+                stringNamed = false;
+                continue;
+            }
+            const actualName = isComputedPropertyName(name) ? name.expression : name;
+            if (isStringLiteralLike(actualName)) {
+                nameText = actualName.text;
+                singleQuote &&= !isStringDoubleQuoted(actualName, currentSourceFile);
+                continue;
+            }
+            stringNamed = false;
+            singleQuote = false;
+            if (isIdentifier(actualName)) {
+                if (actualName !== name) return undefined;
+                nameText = unescapeLeadingUnderscores(actualName.escapedText);
+            }
+            else if (isNumericLiteral(actualName)) {
+                nameText = actualName.text;
+            }
+            else if (
+                isPrefixUnaryExpression(actualName)
+                && isNumericLiteral(actualName.operand)
+            ) {
+                if (actualName.operator === SyntaxKind.PlusToken) {
+                    nameText = actualName.operand.text;
+                }
+            }
+        }
+
+        if (nameText === undefined) {
+            return undefined;
+        }
+
+        return createPropertyNameNodeForIdentifierOrLiteral(
+            nameText,
+            getEmitScriptTarget(options),
+            singleQuote,
+            stringNamed,
+            isMethod,
+        );
+    }
+
+    // Copied similar function in checker. Maybe a reusable one should be created.

```

Comment:
I do think we should pull one of these out; I'm not sure how similar it is, though.

Reply:
They started out more similar. I don't think it ca actually be resable anymore. Will remove the comment.